Add netconf docker to build-image

hussamsh opened the pull request:
<!--
     Please make sure you've read and understood our contributing guidelines:
     https://github.com/Azure/SONiC/blob/gh-pages/CONTRIBUTING.md

     ** Make sure all your commits include a signature generated with `git commit -s` **

     If this is a bug fix, make sure your description includes "fixes #xxxx", or
     "closes #xxxx" or "resolves #xxxx"

     Please provide the following information:
-->

#### Why I did it
I wanted to add NETCONF server support for configuring SONiC switches. Similar to the work done on RESTCONF server.

#### How I did it
First, the changes themselves are on the [`Official sonic netconf repositroy`](https://github.com/sonic-net/sonic-netconf-server).
This work focuses on integrating the netconf server in build-image processes. I updated relevant build scripts , dockerfiles and git modules to include the netconf server in the final image.

#### How to verify it
1) Normally build the image
2) Install new image with netconf server on a switch. After switch stabilizes inspect the docker to see the netconf-server docker image is up.
3) Test NETCONF connection using `ssh <username>@<ip> -p 830 -s netconf`

<!--
If PR needs to be backported, then the PR must be tested against the base branch and the earliest backport release branch and provide tested image version on these two branches. For example, if the PR is requested for master, 202211 and 202012, then the requester needs to provide test results on master and 202012.
-->

#### Description for the changelog
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->

This change adds support for a RFC compliant NETCONF server for SONiC switches.

<!--
 Ensure to add label/tag for the feature raised. example - PR#2174 under sonic-utilities repo. where, Generic Config and Update feature has been labelled as GCU.
-->

Implements sonic-net/SONiC#1466
Implements sonic-net/SONiC#1498
Uses sonic-net/sonic-netconf-server#2

hussamsh commented:
Hello @zhangyanzhao @aseaudi 

This is the remaining part of the NETCONF server implementation, integration in the sonic-build-image.

For your info

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

hussamsh commented:
@lguohan 
Work on this is almost done, can we get some feedback if we can merge this feature ?

EmmaLin11 commented:
Hi @hussamsh,

I'm testing this PR locally and I've encountered a build failure. The error log points to a dependency issue within the Go modules. It looks like github.com/Azure/sonic-mgmt-common requires github.com/pkg/profile@v1.7.0, but the necessary go.sum entry is missing, which prevents the build from completing.

Here is the full error log for your reference:
```
ubuntu$ make target/docker-sonic-netconf-server.gz
...
go: downloading github.com/godbus/dbus/v5 v5.1.0
go: downloading google.golang.org/genproto v0.0.0-20200319113533-08878b785e9c
go: github.com/Azure/sonic-mgmt-common@v0.0.0-20220120155510-515652700481 requires
        github.com/pkg/profile@v1.7.0: missing go.sum entry; to add it:
        go mod download github.com/pkg/profile
make[2]: *** [Makefile:43: vendor/.done] Error 1
make[2]: Leaving directory '/sonic/src/sonic-netconf-server'
dh_auto_build: error: make -j1 returned exit code 2
make[1]: *** [debian/rules:3: build] Error 25
make[1]: Leaving directory '/sonic/src/sonic-netconf-server'
dpkg-buildpackage: error: debian/rules build subprocess returned exit status 2
[  FAIL LOG END  ] [ target/debs/bookworm/sonic-netconf-server_1.0-01_amd64.deb ]
make: *** [slave.mk:790: target/debs/bookworm/sonic-netconf-server_1.0-01_amd64.deb] Error 1
Makefile.work:611: recipe for target 'target/docker-sonic-netconf-server.gz' failed
make[1]: *** [target/docker-sonic-netconf-server.gz] Error 2
make[1]: Leaving directory '/home/ubuntu/SONiC-GITHUB-Build-Harden-202411.0/sonic-buildimage'
Makefile:47: recipe for target 'target/docker-sonic-netconf-server.gz' failed
make: *** [target/docker-sonic-netconf-server.gz] Error 2
```

As a quick test, I manually ran go mod download github.com/pkg/profile which pulled profile@v1.7.0, and the build proceeded without this error.

Would it be possible to update the go.mod and go.sum files in this PR?

msosyak commented:
The HLD https://github.com/sonic-net/SONiC/pull/1466  states that NECONF SERVR will support the following operations:
```
###### 3.2.2.11.9 NETCONF Operations
```
```
Netconf server supports the following operations:
* <get> (RFC 6241 Section 7.7)
* <get-config> (RFC 6241 Section 7.1)
* <copy-config> (RFC 6241 Section 7.3)
* <lock> (RFC 6241 Section 7.5)
* <unlock> (RFC 6241 Section 7.6)
* <close-session> (RFC 6241 Section 7.8)
* <kill-session> (RFC 6241 Section 7.9)
* <get-schema> (RFC 6022 Section 3.1)
```
But in the implementation https://github.com/sonic-net/sonic-netconf-server/blob/master/netconf/server/handler.go#L167 only `get` and `close-session` are mentioned. 

Will the full implementation, as described in the HLD, be contributed to the community? If so, when?

Or do I miss something and the `urn:ietf:params:netconf:capability:writable-running:1.0` is implemented in some different way?

EmmaLin11 commented:
Hi @hussamsh,

Thanks for the quick update and for looking into this!

That's great to hear you'll be updating the PR to the latest sonic-netconf-server commit. This should definitely resolve the runtime panic we observed with the older version.

On a related note regarding sonic-mgmt-common, I've observed that the later commits of sonic-netconf-server seem to call translib.GetYanglibInfo. This function might not be exported in the current sonic-mgmt-common master branch. So, if we move to the latest sonic-netconf-server, it looks like another PR might be needed in sonic-mgmt-common to export translib.GetYanglibInfo to ensure proper build and functionality.

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

mssonicbld commented:
/azp run Azure.sonic-buildimage

hussamsh committed:
Add and configure sonic-netconf-server support

- Add sonic-netconf-server Git submodule
- Add Dockerfile and related configuration for netconf container
- Add systemd service and timer files for netconf server
- Add Makefile rules and dependencies for sonic-netconf-server

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

hussamsh committed:
Remove init_cfg.json from protected files

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

hussamsh committed:
Update netconf module to include leaf-list parsing

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

qiluo-msft commented on rules/docker-sonic-netconf-server.mk:
```diff
@@ -0,0 +1,38 @@
+endif
+
+$(DOCKER_NETCONF_SERVER)_CONTAINER_NAME = netconf-server
+$(DOCKER_NETCONF_SERVER)_RUN_OPT += -v /etc/sonic:/etc/sonic:rw
```
This is security concern. why you need it as rw?

hussamsh replied:
When the NETCONF first starts it needs to generate SSH keys for authentication. One edge case we found in operation is that if the NETCONF server restarts for any reason it would regenerate the key and all the existing clients using NETCONF will have to re-add it locally. That is why when the server starts the first time it checks for the `/etc/sonic/netconf-key.pub` to see if there an existing key, if so use it, if not generate a new key and add there for future use. This makes sure that any client using NETCONF server needs to add the key once across any reboot that might occur.

qiluo-msft commented on rules/docker-sonic-netconf-server.mk:
```diff
@@ -0,0 +1,38 @@
+
+$(DOCKER_NETCONF_SERVER)_CONTAINER_NAME = netconf-server
+$(DOCKER_NETCONF_SERVER)_RUN_OPT += -v /etc/sonic:/etc/sonic:rw
+$(DOCKER_NETCONF_SERVER)_RUN_OPT += -v /etc:/host_etc:ro
```
This is security concern. why you need it at all?

hussamsh replied:
It is needed so because the default authentication method is local authentication - meaning username / password of the host - so this needs to be mounted inside the docker such that local authentication could happen ? Can you suggest a better approach for this ?

qiluo-msft replied:
A compromised or buggy NETCONF container would expose info, such as /etc/shadow. Could you use more secure authentication, such as cert based authen?

qiluo-msft replied:
`/etc/sonic:rw` is also weird, why need to modify this folder? some subfolder has credential, could you not to mount all of them?

qiluo-msft commented on rules/docker-sonic-netconf-server.mk:
```diff
@@ -0,0 +1,38 @@
+$(DOCKER_NETCONF_SERVER)_CONTAINER_NAME = netconf-server
+$(DOCKER_NETCONF_SERVER)_RUN_OPT += -v /etc/sonic:/etc/sonic:rw
+$(DOCKER_NETCONF_SERVER)_RUN_OPT += -v /etc:/host_etc:ro
+$(DOCKER_NETCONF_SERVER)_RUN_OPT += -v /var/run/dbus:/var/run/dbus:rw
```
Could you explain which functionality require `/var/run/dbus`?

hussamsh replied:
The RPC functionality calls functions exposed by the dbus - defined in the `sonic-host-services` .  This option opens communication with dbus on the host.

qiluo-msft commented on rules/config:
```diff
@@ -183,6 +183,9 @@ INCLUDE_SFLOW = y
 INCLUDE_MGMT_FRAMEWORK = y
 
+# INCLUDE_NETCONF_SERVER - build docker-sonic-netconf server for NETCONF support
+INCLUDE_NETCONF_SERVER = y
```
Could it defaulto to n?

hussamsh replied:
yes of course will update this to default `n`.

hussamsh commented on .github/workflows/protect-file.yml:
```diff
@@ -24,7 +24,7 @@ jobs:
       - name: Check protected files
         run: |
           echo "Starting file check step"
-          protected_files=('src/sonic-config-engine/minigraph.py' 'files/build_templates/init_cfg.json.j2')
+          protected_files=('src/sonic-config-engine/minigraph.py')
           base_branch=$(git rev-parse origin/${GITHUB_BASE_REF})
```
the `init_cfg.json.j2` needs to be updated so that the netconf feature can be added. I can re-add the protected file restriction after netconf is added.

hussamsh committed:
fix startup script arg handling and timer description

- Use `// empty` in jq expressions to avoid appending literal "null"
  for absent port/log_level fields in NETCONF_SERVER_ARGS
- Fix netconf-server.timer description to reference the correct container
- Change INCLUDE_NETCONF_SERVER default to `?= n` to allow override
  without forcing the value

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

hussamsh commented:
Hi @qiluo-msft , I just wanted to follow up on the comments you left. if you could take another look so I can move forward with the PR. Thanks a lot!

qiluo-msft commented on dockers/docker-sonic-netconf-server/netconf-server.sh:
```diff
@@ -0,0 +1,34 @@
+
+export CVL_SCHEMA_PATH=/usr/sbin/schema
+
+exec /usr/sbin/netconf_server ${NETCONF_SERVER_ARGS}
```
The unquoted expansion of NETCONF_SERVER_ARGS undergoes word splitting. A CONFIG_DB writer can supply `port: "830 --any-flag value"` and inject arbitrary netconf_server flags.

qiluo-msft commented on files/build_templates/init_cfg.json.j2:
```diff
@@ -78,6 +78,7 @@
 {%- if sonic_asic_platform == "vs" %}{% do features.append(("gbsyncd", "enabled", false, "enabled")) %}{% endif %}
 {%- if include_iccpd == "y" %}{% do features.append(("iccpd", "disabled", false, "enabled")) %}{% endif %}
 {%- if include_mgmt_framework == "y" %}{% do features.append(("mgmt-framework", "enabled", true, "enabled")) %}{% endif %}
+{%- if include_netconf_server == "y" %}{% do features.append(("netconf-server", "enabled", true, "enabled")) %}{% endif %}
```
Do not change image default behavior. You can use /etc/sonic/config_db.json FEATURE table.

qiluo-msft commented on files/build_templates/netconf-server.timer:
```diff
@@ -0,0 +1,11 @@
+PartOf=netconf-server.service
+
+[Timer]
+OnUnitActiveSec=0 sec
```
Recommend removing `OnUnitActiveSec`, which means tight restart loops on service failure

qiluo-msft commented on files/build_templates/netconf-server.service.j2:
```diff
@@ -1,17 +1,13 @@
-Description=Redfish container
+Description=Netconf Server container
+Requires=database.service
+After=database.service swss.service syncd.service
```
syncd.service doesn't exist on chassis supervisor cards / vs-platform (only on line cards/single-asic). The new feature only depends on database, right?

qiluo-msft commented on files/build_templates/netconf-server.service.j2:
```diff
@@ -1,17 +1,13 @@
 After=sonic.target
-StartLimitIntervalSec=1200
-StartLimitBurst=3
+Before=ntp-config.service
```
help me understand why before ntp?

hussamsh replied:
No specific reason, removed it.

hussamsh committed:
Address netconf-server review feedback

- Pass netconf_server args as a quoted bash array (prevent flag injection)
- Default the netconf-server feature to disabled in init_cfg
- Drop OnUnitActiveSec from the timer to avoid tight restart loops
- Depend only on database.service and drop Before=ntp-config.service

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

hussamsh committed:
Remove unused /etc:/host_etc bind mount from netconf-server

Auth uses an SSH loopback to the host sshd, not host /etc files, so the
mount only exposed /etc/shadow with no functional purpose.

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

securely1g reviewed (commented):
Review on behalf of @lguohan:

This PR has been open since July 2024 (over a year). Questions:

1. Is the netconf server still being actively pursued for SONiC? The community has largely standardized on gNMI.
2. The PR modifies .gitmodules and adds a new docker — what is the maintenance commitment?
3. Has this been rebased recently? Given the age, there are likely significant merge conflicts.

Please provide a status update on relevance of this PR.

nakano-omw commented on dockers/docker-sonic-netconf-server/Dockerfile.j2:
```diff
@@ -0,0 +1,34 @@
+debs/{{ deb }}{{' '}}
+{%- endfor %}
+
+COPY ["start.sh", "netconf-server.sh", "/usr/bin/"]
```
Heads-up @hussamsh — while bringing up the netconf-server container from this PR I hit a blocker: the container13:02:23 [49/1956]
 netconf_server never reaches its :830 listen socket, because the two helper scripts are installed without an executable bit.

 This is not specific to our downstream patches. It reproduces on a pristine baseline = this PR's head (070da4f96) + INCLUDE_NETCONF_SERVER=y, no other changes.

```
 $ docker exec netconf-server supervisorctl status
 netconf-server                   STOPPED   Not started
start                            FATAL     command at '/usr/bin/start.sh' is not executable

 $ docker exec netconf-server ls -la /usr/bin/start.sh /usr/bin/netconf-server.sh
 -rw-rw-r-- 1 root root 1037 /usr/bin/netconf-server.sh
-rw-rw-r-- 1 root root  121 /usr/bin/start.sh

$ docker logs netconf-server 2>&1 | grep "not executable"
 INFO spawnerr: command at '/usr/bin/start.sh' is not executable
 INFO gave up: start entered FATAL state, too many start retries too quickly
```

Root cause
dockers/docker-sonic-netconf-server/Dockerfile.j2 does:
```
COPY ["start.sh", "netconf-server.sh", "/usr/bin/"]
```

Both scripts are tracked in Git as mode 100644. docker build preserves the source mode on COPY, so the in-container files inherit
-rw-rw-r--, and supervisord (which checks the exec bit before spawning) refuses them.

 Suggested fix — one line after the COPY:
```
COPY ["start.sh", "netconf-server.sh", "/usr/bin/"]
RUN chmod 0755 /usr/bin/start.sh /usr/bin/netconf-server.sh
COPY ["mgmt_vars.j2", "/usr/share/sonic/templates/"]
```

This makes the in-container mode deterministic regardless of the source-tree file mode — the same pattern already used by other
SONiC dockers. After this, the container comes up on its own:

```
$ docker exec netconf-server ss -tlnp | grep 830
LISTEN 0 512 *:830 *:* users:(("netconf_server",pid=127,fd=7))
$ docker exec netconf-server ls /etc/sonic/netconf-key*
/etc/sonic/netconf-key  /etc/sonic/netconf-key.pub
```

hussamsh committed:
Set netconf scripts executable

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

qiluo-msft reviewed (commented):
A few follow-ups — mostly on the `sonic-netconf-server` submodule (host key + auth), plus inline comments on the build-image files.

**Host key handling (`sonic-netconf-server`, `netconf/main/main.go`):**
- Permissions: the private key is written via `os.Create(...)` → effectively `0644` (world-readable). It must be `0600` (public key `0644`).
- Algorithm: currently hardcoded RSA-1024. Please make the algorithm/size configurable (e.g. a flag fed from `mgmt_vars.j2`, like `port`/`log_level`) with a strong default — ed25519 or RSA ≥ 2048.
- Path: no need to make it runtime-configurable, but please relocate the hardcoded `/etc/sonic/netconf-key` path into a dedicated subdir (e.g. `/etc/sonic/netconf/`) so the host mount can be scoped (see the `.mk` comment).
- Re: reusing the host's main sshd key — I'd avoid it. Mounting the host's primary `ssh_host_*_key` into this container widens the blast radius to all SSH on the box if the container is compromised. A dedicated, well-permissioned key is better.

**Auth model (`sonic-netconf-server`, `netconf/server/auth.go`):**
- Reachability: `Authenticate()` does `ssh.Dial("127.0.0.1:22", ...)`, but I don't see `--network host` in the run options or an sshd in `supervisord.conf`. How does this loopback reach the host sshd? Please clarify (and if it relies on host networking, call that out as a separate consideration).
- Side effects: authenticating by spawning a real SSH login means every NETCONF auth shows up as a localhost sshd login in the audit log plus added latency, and `Authorize()`/`Account()` both return `true` (no RBAC — any valid host user gets full `get`). Worth a follow-up plan even if not blocking v1.

qiluo-msft commented on rules/docker-sonic-netconf-server.mk:
```diff
@@ -0,0 +1,37 @@
+endif
+
+$(DOCKER_NETCONF_SERVER)_CONTAINER_NAME = netconf-server
+$(DOCKER_NETCONF_SERVER)_RUN_OPT += -v /etc/sonic:/etc/sonic:rw
```
`-v /etc/sonic:/etc/sonic:rw` is too broad — `/etc/sonic` holds credentials, and a compromised NETCONF container would get write access to all of it. The only thing needing persistence here is the SSH host key. Please scope this to a dedicated subdir, e.g. `-v /etc/sonic/netconf:/etc/sonic/netconf:rw`.

Note this is coupled with the server binary: the key path is currently hardcoded to `/etc/sonic/netconf-key` in `sonic-netconf-server` (`netconf/main/main.go`), so scoping the mount requires relocating that hardcoded path into the same subdir (it doesn't need to be runtime-configurable — just moved), then bumping the submodule pin.

qiluo-msft commented on dockers/docker-sonic-netconf-server/mgmt_vars.j2:
```diff
@@ -0,0 +1,5 @@
+{
+    "netconf_server": {% if NETCONF_SERVER is defined and "default" in NETCONF_SERVER.keys() %}{{ NETCONF_SERVER['default'] }}{% else %}""{% endif %},
+    "x509" : {% if "x509" in DEVICE_METADATA.keys() %}{{ DEVICE_METADATA["x509"] }}{% else %}""{% endif %},
```
`x509` and `tacacs_plus` are rendered here but are not consumed anywhere in the server binary at the pinned commit (only `port` and `log_level` are read by `netconf-server.sh`). This implies a security capability (cert auth / TACACS) that doesn't exist. Please either wire them up or remove them until they're implemented, so the config isn't misleading.

qiluo-msft commented on files/build_templates/init_cfg.json.j2:
```diff
@@ -78,6 +78,7 @@
 {%- if sonic_asic_platform == "vs" %}{% do features.append(("gbsyncd", "enabled", false, "enabled")) %}{% endif %}
 {%- if include_iccpd == "y" %}{% do features.append(("iccpd", "disabled", false, "enabled")) %}{% endif %}
 {%- if include_mgmt_framework == "y" %}{% do features.append(("mgmt-framework", "enabled", true, "enabled")) %}{% endif %}
+{%- if include_netconf_server == "y" %}{% do features.append(("netconf-server", "disabled", true, "enabled")) %}{% endif %}
```
No need to add `netconf-server` to `init_cfg.json.j2` — please revert this. The feature can be enabled at runtime via the `FEATURE` table in `/etc/sonic/config_db.json` without baking it into the image default.

Since `init_cfg.json.j2` then no longer changes, the removal of it from `.github/workflows/protect-file.yml` is also unnecessary — please revert that too and keep the guardrail intact (see the comment there).

qiluo-msft commented on .github/workflows/protect-file.yml:
```diff
@@ -24,7 +24,7 @@ jobs:
         run: |
           echo "Starting file check step"
-          protected_files=('src/sonic-config-engine/minigraph.py' 'files/build_templates/init_cfg.json.j2')
+          protected_files=('src/sonic-config-engine/minigraph.py')
```
Removing `init_cfg.json.j2` from the protected list drops a review guardrail to land this change. Since the `init_cfg.json.j2` edit isn't needed (configure the feature via `/etc/sonic/config_db.json` instead), please revert this line and keep the protection intact.

hussamsh committed:
Update signed NETCONF server submodule

Point sonic-netconf-server at the signed-off host key hardening commits used by the submodule PR.

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

hussamsh committed:
Revert NETCONF init config entry

Do not add netconf-server to init_cfg.json.j2 by default and restore the protected-file guardrail for init_cfg.json.j2.

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

hussamsh committed:
Configure NETCONF host key algorithm

Pass NETCONF host key algorithm and RSA size settings from container config to the server, defaulting to ed25519 and RSA 3072 bits.

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

hussamsh committed:
Remove unused NETCONF auth vars

Stop rendering x509 and tacacs_plus in the NETCONF container template because the startup script and pinned server binary do not consume those fields.

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

hussamsh committed:
Scope netconf-server host key mount

Mount only /etc/sonic/netconf into the NETCONF container for persistent host keys and bump sonic-netconf-server to generate keys in that dedicated directory.

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

hussamsh committed:
Update NETCONF server key permissions

Bump sonic-netconf-server to create generated NETCONF SSH host keys with restricted file permissions.

Signed-off-by: Hossam Hassan <hossam4.hassan@orange.com>

qiluo-msft commented on rules/docker-sonic-netconf-server.mk:
```diff
@@ -0,0 +1,37 @@
+endif
+
+$(DOCKER_NETCONF_SERVER)_CONTAINER_NAME = netconf-server
+$(DOCKER_NETCONF_SERVER)_RUN_OPT += -v /etc/sonic/netconf:/etc/sonic/netconf:rw
```
the mount was narrowed from the full `/etc/sonic` to `/etc/sonic/netconf` — that's better. follow-up: does the server actually write to this directory, or is it read-only config? if the container only reads its config from here, `:ro` is sufficient and limits blast radius from a compromised container.

qiluo-msft commented on dockers/docker-sonic-netconf-server/netconf-server.sh:
```diff
@@ -0,0 +1,41 @@
+[ -n "$HOST_KEY_ALGORITHM" ] && NETCONF_SERVER_ARGS+=("-host_key_algorithm" "$HOST_KEY_ALGORITHM")
+[ -n "$HOST_KEY_RSA_BITS"  ] && NETCONF_SERVER_ARGS+=("-host_key_rsa_bits" "$HOST_KEY_RSA_BITS")
+
+echo "NETCONF_SERVER_ARGS = ${NETCONF_SERVER_ARGS[*]}"
```
nit: `${NETCONF_SERVER_ARGS[*]}` unquoted — word-splits on spaces in any element. harmless here (just logging), but inconsistent with the correctly-quoted `"${NETCONF_SERVER_ARGS[@]}"` on line 41. use `"${NETCONF_SERVER_ARGS[*]}"` or `"${NETCONF_SERVER_ARGS[@]}"`.

qiluo-msft commented on rules/docker-sonic-netconf-server.mk:
```diff
@@ -0,0 +1,37 @@
+$(DOCKER_NETCONF_SERVER)_DBG_DEPENDS = $($(DOCKER_CONFIG_ENGINE_BOOKWORM)_DBG_DEPENDS)
+$(DOCKER_NETCONF_SERVER)_DBG_DEPENDS += $(SONIC_NETCONF_SERVER_DBG)
+
+SONIC_DOCKER_IMAGES += $(DOCKER_NETCONF_SERVER)
```
`SONIC_DOCKER_IMAGES += $(DOCKER_NETCONF_SERVER)` runs unconditionally, but `$(SONIC_NETCONF_SERVER)` (which the docker depends on, line 10) is only defined when `INCLUDE_NETCONF_SERVER = y` (it's wrapped in an `ifeq` in `sonic-netconf-server.mk`). when `INCLUDE_NETCONF_SERVER = n`, the docker is added to the build list but its dependent deb doesn't exist → the docker build will fail. either wrap this line in `ifeq ($(INCLUDE_NETCONF_SERVER), y)` alongside the INSTALL lines, or make the package always build (like `SONIC_MGMT_COMMON`).
